### PR TITLE
Try to bypass query with only catalog tables

### PR DESCRIPTION
**Pull request by RMTT**
Try to bypass query with only catalog tables.

We need to bypass the query with only catalog tables in resource groups in some situations.

I just walk the parse tree and find `RangeVar` to compare it's `schemaname` with `pg_catalog`, it's just a few cases which be considered now.

### CODE REVIEW

File: src/backend/utils/resgroup/resgroup.c [hunk 1]
```diff
@@ -3677,6 +3677,26 @@ groupWaitQueueFind(ResGroupData *group, const PGPROC *proc)
...
 
+static bool checkBypassWalker(Node *node, void *context) {
+	ListCell *cell;
+	bool *bypass = context;
+
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, RangeVar)) {
+		foreach(cell, (List *)node) {
```
Comment (kainwen) [inline src/backend/utils/resgroup/resgroup.c:3688]: When it is `RangeVar`, we enter this scope, why we loop it as `List`?

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -3736,7 +3756,11 @@ shouldBypassQuery(const char *query_string)
 	{
 		parsetree = (Node *) lfirst(parsetree_item);
 
-		if (nodeTag(parsetree) != T_VariableSetStmt &&
+		if (IsA(parsetree, SelectStmt)){
+			raw_expression_tree_walker(parsetree, checkBypassWalker, &bypass);
```
Comment (kainwen) [inline src/backend/utils/resgroup/resgroup.c:3760]: Format issue, and personally I recommend adding a wrapper function `is_total_catlog_sql` (find a better name yourself please) and invoke it here.

File: src/backend/utils/resgroup/resgroup.c [hunk 3]
```diff
@@ -3736,7 +3756,11 @@ shouldBypassQuery(const char *query_string)
...
+			if (bypass == false)
+				break;
+		}else if (nodeTag(parsetree) != T_VariableSetStmt &&
```
Comment (kainwen) [inline src/backend/utils/resgroup/resgroup.c:3763]: I do not think we should use `nodeTag` here, not this PR's fault, let's take the chance to refactor the code.

File: src/backend/utils/resgroup/resgroup.c (hunk 1 above)
Comment (RMTT) [reply src/backend/utils/resgroup/resgroup.c:3688]: I forget remove the `for` loop, good suggestion!

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (RMTT) [reply src/backend/utils/resgroup/resgroup.c:3760]: I think it too easy to create a function.

**Comment by ashwinstar** [comment]
- Nice to see movement for enabling this capability, looking forward
- Interesting to see we are starting with 6X_STABLE. I think development should happen on main branch first.
- we will need to make this opt-in behavior via GUC or some other mechanism. Would be good to provide those details in PR description and commit message how user facing aspect will look like
- Also, need to think if we wish to provide bypass capability for RG level or full Cluster level as whole
- Plus, probably only applies on coordinator node I guess and doesn't make sense for segment nodes (or given concurrency control only happens on coordinator defacto anyways will only get triggered for coordinator)

**Pull request closed without merging**